### PR TITLE
[Quant Tool] Prevent int32 quantized bias from clipping by adjusting the weight's scale

### PR DESCRIPTION
**Pull request by adrianlizarraga**
### Description
Fixes scenario in which a bias input quantized to int32 has a scale that is too small. A bias with a scale that is smaller than a certain threshold will overflow the range of an `int32` when quantized, which significantly decreases accuracy.

Credit to @yihonglyu for finding out about this issue and the fix.

### Motivation and Context
Consider the following Convolution with very small weights and a constant bias input of `[5, -4.5]`.
image

The QDQ quantizer first computes the following quantization scale for `input_0` and `weight`:
- `input_0`: scale=0.5
- `weight`: scale=7.843e-10 **[really small]**

The QDQ quantizer then computes the bias input's scale as follows:
```
bias_scale = input_0_scale * weight_0_scale = 0.5 * 7.843e-10 = 3.9215686274509805e-11
```

This `bias_scale` is too small. Before this PR, the QDQ quantizer would quantize the f32 bias with this `bias_scale`:
```
bias_quant = round(bias_f32 / bias_scale) =  round([5.0/bias_scale, -4.5/bias_scale]) = [127500000000, -114750000000]
```
These quantized bias values exceed the range of int32, and so are clipped to [int32.min(), int32.max()], which is very inaccurate.

#### New approach
This PR increases the `weight_0_scale` by the necessary amount to ensure that `bias_scale` (which equals `weight_0_scale * input_0_scale`) is appropriate for the int32 quantization type.

The smallest valid bias scale is given by the normal scale formula: 
`bias_smallest_valid_scale = (bias_f32_max - bias_f32_min) / (int32_max - int32_min)`

Then, we compute the candidate bias scale:
`bias_scale_candidate = input_0_scale * weight_0_scale`

If the candidate scale is smaller than the smallest valid scale, we increase the `weight_0_scale` by the necessary ratio:
```python
if bias_scale_candidate < bias_smallest_valid_scale:
    ratio = bias_smallest_valid_scale / bias_scale_candidate
    weight_0_scale = ratio * weight_0_scale
```

Then, we recompute the final bias scale:
```python
bias_scale = input_0_scale * weight_0_scale
```

#### Impact on accuracy
Here's the above model's quantized output compared to the f32 (ground-truth) output.
- Before PR: 
  - f32 model output[0]: **5.0f**
  - qdq model output[0]: **0.075**
  - SNR: 0.1369 (higher is better)
- After PR:
  - f32 model output[0]: **5.0f**
  - qdq model output[0]: **4.992**
  - SNR: 55.656 (higher is better)

### CODE REVIEW

**Comment by adrianlizarraga** [comment]
Found an oversight. Converting back to draft.

**Comment by adrianlizarraga** [comment]
Ready for review

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 1]
```diff
@@ -380,6 +556,8 @@ def quantize_model(self):
                         self.tensor_to_its_receiving_nodes[tensor_name] = []
                     self.tensor_to_its_receiving_nodes[tensor_name].append(node)
 
+        self.initializer_quant_params = self._calc_initializer_quant_params()
+        self._adjust_weight_quant_params_for_bias_tensors()
         self._quantize_normal_tensors()
         self._quantize_sharing_param_tensors()
         if self.quantize_bias:
```
Comment (adrianlizarraga) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:563]: **Note**: This is the central model quantization loop. Previously, we computed scale/zp _and_ quantized initializers at the same time within `self._quantize_normal_tensors()`. Now, we separate the computation of the initializer's scale/zp from the actual quantization. This allows us to modify the quantization parameters (e.g., to adjust them for a bias) separately.

Additionally, as the code grew, the logic for determining the quantization type/symmetry/range for initializers was scattered throughout various functions and files. Now, all of the logic for determining the scale/zp of initializers is consolidated into a single function: `self._calc_initializer_quant_params()`.

**Review by xadupre — APPROVED** [review]

**Comment by snnn** [comment]
This PR has been cherry-picked into the `rel-1.20.1` branch in PR #22785. Removing the `release:1.20.1` label.

**Pull request merged**